src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/SnapshotTracker.kt

```diff
@@ -0,0 +1,231 @@
+    }
+
+    fun trackField(obj: Any?, className: String, fieldName: String) {
+        if (obj != null && obj !in trackedObjects) return
+
+        val clazz: Class<*> = Class.forName(className).let {
+            if (obj != null) it
+            else it.findField(fieldName).declaringClass
+        }
+        val field = clazz.findField(fieldName)
+        val readResult = runCatching { readFieldViaUnsafe(obj, field) }
+
+        if (readResult.isSuccess) {
+            val fieldValue = readResult.getOrNull()
+            trackSingleField(obj, clazz, field, fieldValue) {
+                if (shouldTrackEnergetically(fieldValue)) {
+                    trackHierarchy(fieldValue!!)
+                }
+            }
+        }
+    }
+
+    fun trackArrayCell(array: Any, index: Int) {
+        if (array !in trackedObjects) return
+
+        val readResult = runCatching { readArrayElementViaUnsafe(array, index) }
+
+        if (readResult.isSuccess) {
+            val elementValue = readResult.getOrNull()
+            trackSingleArrayCell(array, index, elementValue) {
+                if (shouldTrackEnergetically(elementValue)) {
+                    trackHierarchy(elementValue!!)
+                }
+            }
+        }
+    }
+
+    fun trackObjects(objs: Array<Any?>) {
+        // in case this works too slowly, an optimization could be used
+        // see https://github.com/JetBrains/lincheck/pull/418/commits/0d708b84dd2bfd5dbfa961549dda128d91dc3a5b#diff-a684b1d7deeda94bbf907418b743ae2c0ec0a129760d3b87d00cdf5adfab56c4R146-R199
+        objs
+            .filter { it != null && it in trackedObjects }
+            .forEach { trackHierarchy(it!!) }
+    }
+
+    fun restoreValues() {
+        val visitedObjects = Collections.newSetFromMap(IdentityHashMap<Any, Boolean>())
+        trackedObjects.keys
+            .filterIsInstance<Class<*>>()
+            .forEach { restoreValues(it, visitedObjects) }
+    }
+
+    private fun trackSingleField(
+        obj: Any?,
+        clazz: Class<*>,
+        field: Field,
+        fieldValue: Any?,
+        callback: (() -> Unit)? = null
+    ) {
+        val nodesList =
+            if (obj != null) trackedObjects[obj]
+            else trackedObjects.getOrPut(clazz) { mutableListOf<MemoryNode>() }
+
+        if (
+            nodesList == null || // parent object is not tracked
+            nodesList
+                .map { it.descriptor }
+                .filterIsInstance<FieldDescriptor>()
+                .any { it.field.name == field.name } // field is already tracked
+        ) return
+
+        val childNode = createMemoryNode(
+            obj,
+            FieldDescriptor(field, getFieldOffset(field)),
+            fieldValue
+        )
+
+        nodesList.add(childNode)
+        if (isTrackableObject(childNode.initialValue)) {
+            trackedObjects.putIfAbsent(childNode.initialValue, mutableListOf<MemoryNode>())
+            callback?.invoke()
```

Comment:
Let's just return `Boolean` instead of passing `callback`, it would simplify the code.

Reply:
done